secure-api-gateway-ig-extensions/src/main/java/com/forgerock/sapi/gateway/jwks/sign/DefaultSapiJwsSigner.java

```diff
@@ -0,0 +1,147 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.forgerock.sapi.gateway.jwks.sign;
+
+import static org.forgerock.openig.secrets.SecretsProviderHeaplet.secretsProvider;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.forgerock.json.jose.builders.JwsHeaderBuilder;
+import org.forgerock.json.jose.builders.JwtBuilderFactory;
+import org.forgerock.json.jose.jws.JwsAlgorithm;
+import org.forgerock.json.jose.jws.SigningManager;
+import org.forgerock.json.jose.jws.handlers.SigningHandler;
+import org.forgerock.json.jose.jwt.JwtClaimsSet;
+import org.forgerock.openig.heap.GenericHeaplet;
+import org.forgerock.openig.heap.HeapException;
+import org.forgerock.secrets.Purpose;
+import org.forgerock.secrets.SecretsProvider;
+import org.forgerock.secrets.keys.SigningKey;
+import org.forgerock.util.promise.Promise;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Default implementation of {@link SapiJwsSigner}
+ * <p>
+ * A default JWS signer. <br/>
+ * Provides signing functionality for use by the filters and/or the IG scripts<br/>
+ * @see DefaultSapiJwsSigner.Heaplet
+ */
+public class DefaultSapiJwsSigner implements SapiJwsSigner {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultSapiJwsSigner.class);
+    private final SigningManager signingManager;
+    private final Purpose<SigningKey> signingKeyPurpose;
+    private final String algorithm;
+    private final String kid;
+
+    public DefaultSapiJwsSigner(
+            SecretsProvider secretsProvider,
+            String signingKeyId,
+            String kid,
+            String algorithm
+    ) {
+        this.signingManager = new SigningManager(secretsProvider);
+        this.kid = kid;
+        this.algorithm = algorithm;
+        this.signingKeyPurpose = Purpose.purpose(signingKeyId, SigningKey.class);
+    }
+
+    @Override
+    public Promise<String, SapiJwsSignerException> sign(
+            final Map<String, Object> payload,
+            final Map<String, Object> criticalHeaderClaims
+    ) {
+        return signingManager.newSigningHandler(signingKeyPurpose)
+                .then(signingHandler -> sign(signingHandler, payload, criticalHeaderClaims),
+                        noSuchSecretException -> {
+                            logger.error("Failed to create signingHandler", noSuchSecretException);
+                            throw new SapiJwsSignerException(noSuchSecretException.getMessage());
```

Comment:
Generally speaking, you want to pass the Throwable cause to the new exception that you are creating otherwise you lose information. Also add some extra information for context, what has caused us to get this exception

e.g.
```
new SapigJwsSignerException("Failed to create signingHandler, noSuchSecretException)
```

Reply:
Done